packages/amplify-graphql-auth-transformer/src/vtl-generator/ddb/resolvers/helpers.ts

```diff
@@ -145,30 +143,54 @@ export const lambdaExpression = (roles: Array<RoleDefinition>): Expression =>
     compoundExpression([...(roles.length > 0 ? [set(ref(IS_AUTHORIZED_FLAG), bool(true))] : [])]),
   );
 
+export type IamExpressionOptions = {
+  roles: Array<RoleDefinition>;
+  adminRolesEnabled: boolean;
+  hasIdentityPoolId: boolean;
+  genericIamAccessEnabled: boolean;
+  fieldName?: string;
+};
+
 /**
  * Creates iam expression helper
  */
-export const iamExpression = (
-  roles: Array<RoleDefinition>,
-  adminRolesEnabled: boolean,
-  hasIdentityPoolId: boolean,
-  fieldName: string = undefined,
-): Expression => {
+export const iamExpression = (options: IamExpressionOptions): Expression => {
   const expression = new Array<Expression>();
   // allow if using an admin role
-  if (adminRolesEnabled) {
-    expression.push(iamAdminRoleCheckExpression(fieldName));
+  if (options.adminRolesEnabled) {
+    expression.push(iamAdminRoleCheckExpression(options.fieldName));
   }
-  if (roles.length > 0) {
-    roles.forEach((role) => {
+  if (options.roles.length > 0) {
+    options.roles.forEach((role) => {
       expression.push(
-        iff(not(ref(IS_AUTHORIZED_FLAG)), iamCheck(role.claim!, set(ref(IS_AUTHORIZED_FLAG), bool(true)), hasIdentityPoolId)),
+        iff(not(ref(IS_AUTHORIZED_FLAG)), iamCheck(role.claim!, set(ref(IS_AUTHORIZED_FLAG), bool(true)), options.hasIdentityPoolId)),
       );
     });
   } else {
     expression.push(ref('util.unauthorized()'));
   }
-  return iff(equals(ref('util.authType()'), str(IAM_AUTH_TYPE)), compoundExpression(expression));
+  return iff(
+    equals(ref('util.authType()'), str(IAM_AUTH_TYPE)),
+    generateIAMAccessCheck(options.genericIamAccessEnabled, compoundExpression(expression)),
+  );
+};
+
+/**
+ * Creates an expression that allows generic IAM access for principals not associated to CognitoIdentityPool.
+ *
+ * Note: This function assumes that caller already checked that auth type is IAM. Should not be used outside of IAM helpers.
+ */
+export const generateIAMAccessCheck = (enableIamAccess: boolean, expression: Expression): Expression => {
+  if (!enableIamAccess) {
+    // No-op if generic IAM access is not enabled.
+    return expression;
+  }
+
+  const isGenericIamAccess = and([
```

Comment:
It's because all these calls are nested under `iamExpression` so we'd be repeating auth type check.
See https://github.com/aws-amplify/amplify-category-api/blob/82ae3aac071d64acbe4727631850ef7ee8f88324/packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/iam-access.test.ts.snap#L645-L654 .

I.e. 
instead of 
```
#if( $util.authType() == \\"IAM Authorization\\" )
  #if( $util.isNull($ctx.identity.cognitoIdentityPoolId) && $util.isNull($ctx.identity.cognitoIdentityId) )
    $util.qr($ctx.stash.put(\\"hasAuth\\", true))
    #set( $isAuthorized = true )
  #else
    #if( $ctx.identity.userArn == $ctx.stash.unauthRole )
      #set( $isAuthorized = true )
    #end
  #end
#end
```

We'd have (double `$util.authType() == \\"IAM Authorization\\"`).
```
#if( $util.authType() == \\"IAM Authorization\\" )
  #if( $util.authType() == \\"IAM Authorization\\" && $util.isNull($ctx.identity.cognitoIdentityPoolId) && $util.isNull($ctx.identity.cognitoIdentityId) )
    $util.qr($ctx.stash.put(\\"hasAuth\\", true))
    #set( $isAuthorized = true )
  #else
    #if( $ctx.identity.userArn == $ctx.stash.unauthRole )
      #set( $isAuthorized = true )
    #end
  #end
#end
```

Reply:
Having said that. It's fine to reuse `isNonCognitoIAMPrincipal` if we don't mind double-check (I guess we don't but looking for confirmation).

Reply:
I see your point. My concern is that someone is going to incorrectly use `generateIAMAccessCheck` outside of a block that has already asserted IAM auth exists. This discussion suggests that we should probably separate this out further, but I'm reluctant to take on any more refactoring as part of this change.

My vote would be that it's OK to duplicate the generated code in the interests of a secure-by-default utility method.

Reply:
changed in https://github.com/aws-amplify/amplify-category-api/pull/2331/commits/7eae99b5dc8e5965842a38ddc41efe4410ba614b